### PR TITLE
Reply rendering

### PR DESCRIPTION
**Pull request by stefanceriu**
The RustSDK has stopped seding us reply fallbacks as Matrix as a whole is moving away from them so we now need to start using the already available `InReplyToDetails`. 

This PR refactors the existing code to extract out the types common to `MessageTimelineItems` and `RepliedToEventDetails` so that we can now bundle `MessageContent` together with both. All message type timeline items now have a `replyDetails` property that they can use in rendering.
It also exposes a new Room method for fetching `.unavailable` event details when a certain reply appears on screen (`RoomTimelineController.processItemAppearance`)

Screenshot 2023-05-08 at 12 14 34
Screenshot 2023-05-08 at 12 14 59

Known issues:
- [ ] Reply details sometimes don't load
- [ ] They are only rendered in textual format for all message content types (e.g. we won't show a thumbnail in the reply bubble)
- [ ] The reply bubble doesn't expand to fill all the available space. Couldn't figure out how to do that, not even with a custom Layout

### CODE REVIEW

**Comment by Hywan** [comment]
👏

File: ElementX/Sources/Screens/RoomScreen/View/Style/TimelineItemBubbledStylerView.swift [hunk 1]
```diff
@@ -123,6 +123,22 @@ struct TimelineItemBubbledStylerView<Content: View>: View {
...
+            VStack(alignment: .leading, spacing: 4.0) {
+                TimelineReplyView(timelineItemReplyDetails: replyDetails)
+                    .background(Color.element.background)
+                    .cornerRadius(8)
+                
+                content()
+            }
+        } else {
+            content()
+        }
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/View/Style/TimelineItemBubbledStylerView.swift:139]: Might be simpler to move the if let inside the `VStack` here so there's only a single `content()` call? An extra `VStack` shouldn't really be a layout concern. (Same comment on the plain styler)

File: ElementX/Sources/Services/Timeline/TimelineItemContent/TimelineItemReplyDetails.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    case image(ImageRoomTimelineItemContent)
+    case notice(NoticeRoomTimelineItemContent)
+    case text(TextRoomTimelineItemContent)
+    case video(VideoRoomTimelineItemContent)
+}
+
+enum TimelineItemReplyDetails: Hashable {
+    case unavailable(eventID: String)
+    case pending(eventID: String)
+    case ready(sender: TimelineItemSender, content: TimelineItemReplyContent)
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/TimelineItemContent/TimelineItemReplyDetails.swift:32]: In fact how about
```suggestion
    case notLoaded(eventID: String)
    case loading(eventID: String)
    case loaded(sender: TimelineItemSender, content: TimelineItemReplyContent)
```

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift [hunk 3]
```diff
@@ -412,4 +442,54 @@ struct RoomTimelineItemFactory: RoomTimelineItemFactoryProtocol {
...
+            case .image(let content):
+                replyContent = .image(buildImageTimelineItemContent(content))
+            case .notice(let content):
+                replyContent = .notice(buildNoticeTimelineItemContent(content))
+            case .text(let content):
+                replyContent = .text(buildTextTimelineItemContent(content))
+            case .video(let content):
+                replyContent = .video(buildVideoTimelineItemContent(content))
+            case .none:
+                replyContent = .text(TextRoomTimelineItemContent(body: "Unkown reply content"))
```
Comment (pixlwave) [inline ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:487]: Should this one be localised too, plus small typo.
```suggestion
                replyContent = .text(TextRoomTimelineItemContent(body: "Unknown reply content"))
```

File: ElementX/Sources/Screens/RoomScreen/View/Replies/TimelineTextReplyView.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    private func attributedString(for sender: TimelineItemSender, body: String, formattedBody: AttributedString?) -> AttributedString {
+        var attributedHeading = AttributedString("\(sender.displayName ?? sender.id)\n")
+        attributedHeading.font = .compound.bodyMD.bold()
+        attributedHeading.foregroundColor = .element.primaryContent
+        
+        var formattedBody = formattedBody ?? AttributedString(body)
+        formattedBody.font = .compound.bodyMD
+        formattedBody.foregroundColor = .element.secondaryContent
+        
+        attributedHeading += formattedBody
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/View/Replies/TimelineTextReplyView.swift:60]: TBH, this should really be a `VStack` with the heading separated from the body (which only needs 2 lines at that point), but I'm happy to look at that in the follow-up with layout fixes.

**Review by pixlwave — APPROVED** [review]
Overall looks good to me.
I think some of the comments on the 1st commit might have got a bit lost in the changes from the subsequent ones, so ping me if its not clear.

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift [hunk 5]
```diff
@@ -282,28 +284,40 @@ class RoomScreenViewModel: RoomScreenViewModelType, RoomScreenViewModelProtocol
...
         case .react:
-            callback?(.displayEmojiPicker(itemID: item.id))
+            callback?(.displayEmojiPicker(itemID: eventTimelineItem.id))
         case .copy:
-            UIPasteboard.general.string = item.body
+            guard let messageTimelineItem = timelineItem as? EventBasedMessageTimelineItemProtocol else {
+                return
+            }
+            
+            UIPasteboard.general.string = messageTimelineItem.body
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:302]: Not sure how we would do that properly considering they're not actually loaded yet, just their thumbnails.

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift [hunk 6]
```diff
@@ -149,186 +148,219 @@ struct RoomTimelineItemFactory: RoomTimelineItemFactoryProtocol {
...
-                                    isEditable: eventItemProxy.isEditable,
-                                    sender: eventItemProxy.sender,
-                                    properties: RoomTimelineItemProperties(isEdited: eventItemProxy.content.asMessage()?.isEdited() ?? false,
-                                                                           reactions: aggregateReactions(eventItemProxy.reactions)))
+        TextRoomTimelineItem(id: eventTimelineItem.id,
+                             timestamp: eventTimelineItem.timestamp.formatted(date: .omitted, time: .shortened),
+                             isOutgoing: isOutgoing,
+                             isEditable: eventTimelineItem.isEditable,
+                             sender: eventTimelineItem.sender,
+                             content: .init(body: "Unknown timeline item"),
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:179]: `eventTimelineItem` not longer has a body property, as we don't know if it's a message or not. I think I'll just go with out `common_unsupported_event`

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift [hunk 7]
```diff
@@ -149,186 +148,219 @@ struct RoomTimelineItemFactory: RoomTimelineItemFactoryProtocol {
...
-    private func buildImageTimelineItemFromMessage(_ eventItemProxy: EventTimelineItemProxy,
-                                                   _ message: MessageTimelineItem<ImageMessageContent>,
+    private func buildImageTimelineItemFromMessage(_ eventTimelineItem: EventTimelineItemProxy,
+                                                   _ messageTimelineItem: Message,
+                                                   _ messageContent: ImageMessageContent,
                                                    _ isOutgoing: Bool) -> RoomTimelineItemProtocol {
+        var thumbnailSource: MediaSourceProxy?
+        if let source = messageContent.info?.thumbnailSource {
+            thumbnailSource = MediaSourceProxy(source: source, mimeType: messageContent.info?.thumbnailInfo?.mimetype)
+        }
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:324]: Yes, good idea 👍

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift (hunk 3 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:487]: Okay for the typo but we shouldn't worry about the other one, I can't see any reason why that message type is optional. I'll ask the rust team.

File: ElementX/Sources/Services/Timeline/TimelineItemContent/TimelineItemReplyDetails.swift (hunk 2 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItemContent/TimelineItemReplyDetails.swift:32]: Looks good to me 👍

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift (hunk 3 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:487]: Actually, even better, just going to return nil here

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:179]: Removed it entirely in the end

File: ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift [hunk 8]
```diff
@@ -149,186 +148,219 @@ struct RoomTimelineItemFactory: RoomTimelineItemFactoryProtocol {
...
+        
+        let htmlBody = messageContent.formatted?.format == .html ? messageContent.formatted?.body : nil
 
         var formattedBody: AttributedString?
-        if let htmlBody = message.htmlBody {
+        if let htmlBody {
             formattedBody = attributedStringBuilder.fromHTML("* \(name) \(htmlBody)")
         } else {
-            formattedBody = attributedStringBuilder.fromPlain("* \(name) \(message.body)")
+            formattedBody = attributedStringBuilder.fromPlain("* \(name) \(messageContent.body)")
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineItems/RoomTimelineItemFactory.swift:397]: Done 👍

**Pull request merged**